### DISCUSSION TITLE
ARX papers: 2020 A comprehensive tool for creating and evaluating privacy-preserving biomedical prediction models

### DISCUSSION THREAD

**Discussion by idhamari** [iaBIH/arx#20] [ARX Papers]
[Eicher, J., Bild, R., Spengler, H. et al. A comprehensive tool for creating and evaluating privacy-preserving biomedical prediction models. BMC Med Inform Decis Mak 20, 29 (2020). https://doi.org/10.1186/s12911-020-1041-3](https://bmcmedinformdecismak.biomedcentral.com/articles/10.1186/s12911-020-1041-3)

**Comment by idhamari** [comment]
* Relative accuracy uses [Brier skill score](https://www.dwd.de/EN/ourservices/seasonals_forecasts/forecast_reliability.html#:~:text=Brier%20Skill%20Score%20(BSS)&text=The%20BSS%20indicates%20the%20degree,BSS%20is%20equal%20to%20one.) which quantifies the relative accuracy of a classification model over some reference accuracy. The paper cites this [reference](https://doi.org/10.1002/qj.709)

              Brier score (BS)  =  $(1/n) sum (y_i - x_i)^2 where x_i: class and  y_i: probablity
              Brier skill score (BSS) = relative accuracy = 1- (Brier_Annonymized / Brier_Original) 
                                          = (Brier_Annonymized - Brier_Original) / Brier_Original

* Calculating models performance is measured using [k-fold cross-validation](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.49.8228). In this process, the records of a dataset are first divided randomly into k partitions of equal size, which are then iteratively analyzed by using each of the k partitions as evaluation and all other partitions as training data. This process yields k results which are combined to derive an overall estimate of the model’s performance.

**Reply by idhamari** [reply]
This paper explains the relationship between classification and anonymization in ARX.

**Reply by idhamari** [reply]
* The goal is to have anonymized data with "somehow" similar classification's accuracy to the original data. 
* If the classification accuracy is too bad then the anonymization process produces "useless"  data. 
* data granularity: the size in which data fields are sub-divided e.g. a postal address can be recorded, with coarse granularity, as a single field or fine granularity as multiple fields. 
* [transformation-models](https://arx.deidentifier.org/overview/transformation-models/) vs [privacy-models](https://arx.deidentifier.org/overview/privacy-criteria/)
* ARX should produce all possible output datasets by applying all possible combinations of generalizations to the input dataset. 
    - For each possible output, two parameters are measured: (1) privacy protection, and (2) data utility. 
    - In practice, it uses pruning strategies and optimization to avoid producing all possible output datasets, see [this](https://www.springerprofessional.de/en/putting-statistical-disclosure-control-into-practice-the-arx-dat/6857990) and [this](https://bmcmedinformdecismak.biomedcentral.com/articles/10.1186/s12911-016-0287-2)
* Privacy models:
   - (1) k-anonymity: protects records from re-identification by requiring that each transformed record is indistinguishable from at least k − 1 other records regarding attributes that could be used in linkage attacks.
  - (2) differential privacy: guarantees that the output of the anonymization procedure is basically independent of the contribution of individual records to the dataset, which protects output data from a wide range of risks
  - (3) game-theoretic: employs an economic perspective on data re-identification attacks and assumes that adversaries will only attempt re-identification in case there is a tangible economic benefit
* Utility models:  models for quantifying the utility of anonymized data.
   - [Iyengar method](https://dl.acm.org/doi/abs/10.1145/775047.775089):  distinguish between the removal of structure and the removal of noise by measuring the heterogeneity of values of class attributes in groups of records that are indistinguishable regarding the specified feature variables. The implementation had several important limitations, which resulted from the compressed internal data representation used by ARX: (1) one class variable,  (2)  class variables must be addressed by a privacy model, (3) no transformations to target variables. Major parts of ARX is rewritten. 
* Transformation models:
   - (1) Random sampling: reduce the certainty of attackers about the correctness of re-identifications. It is also a major building block of differential privacy in ARX.
   - (2) Aggregation: sets of numeric attribute values are transformed into a common aggregated value. 
   - (3) Suppression: values are simply removed from a dataset, which may be applied on the cell-, record- or attribute-level.
   - (4) Masking: individual characters are removed. 
   - (5) Categorization: continuous variables are mapped to categories. 
   - (6) Generalization: attribute values are replaced by less specific values based on user-defined generalization hierarchies. 
* Classification models:
   - (1) Logistic regression: the relationship between the feature variables and the target variable is expressed as a linear model which is transformed using a logarithmic function. 
   - (2) Naïve Bayes: makes strong (hence naïve) assumptions about the independence of the distributions of the feature variables based on Bayes’ theorem. The only dependency is assumed to exist between the target variable and each of the feature variables. Predictions are made by simply calculating the posterior probabilities of each of the classes using the prior probability of the feature vector. 
   - (3) Random forest: belongs ensemble learning methods. The predictions of multiple models are combined into a single prediction. The individual models are decision trees generated from independently sampled training data by selecting a random subset of the features at each split in the learning process.

**Reply by idhamari** [reply]
```diff
- why not use Logistic Regression from [SMILE](https://haifengl.github.io/classification.html)? 
- why not add other strong classifiers e.g. SVM?
```

**Comment by idhamari** [comment]
**Preprocessing training data:**

* K-fold cross-validation: the records of a dataset are first divided randomly into k partitions of equal size, which are then iteratively analyzed by using each of the k partitions as evaluation and all other partitions as training data. This process yields k results which are combined to derive an overall estimate of the model’s performance.
* When classification models are built from anonymized data, it needs to be evaluated 
   - how anonymization has affected classification models' performance. 
   - not clear: 
   
```diff
- This cannot be implemented “naively” by comparing the results of performing k-fold cross-validation on the anonymized data and of performing k-fold cross-validation on input data. 
- A classifier must be built from transformed output data in such a way that the model is able to make predictions based on features that have not been transformed. 
```
 - model can be evaluated using unmodified input data to obtain relative performance estimates. 
 - a preprocessing step transforms a given set of previously unknown features in the same manner in which the anonymized training data has been transformed before passing it to the classifier to make predictions. 
 - it can be seen that in the naive approach two classifiers are built from two different datasets (input and output), evaluated against these datasets, and then their accuracy is compared to derive a relative performance.
 - the second classifier is built from output data but evaluated on (preprocessed) input data to obtain comparable results for both models.
![Screenshot from 2022-03-14 15-44-11](https://user-images.githubusercontent.com/28802516/158196228-0541ab19-7a8f-465b-a3e5-17eca4bdef3b.png)

- the new tool creates privacy-preserving classification models by training them on anonymized data. 
- this results in the challenge that the prediction models created can only be applied to data that has been transformed in the same way as the anonymized training dataset. 
- the resulting prediction models must be able to interpret features from output data as well as input data correctly. 
- this is challenging when the domain of attribute values is not preserved during anonymization, as in these cases, the input contains values that are not present in the output, and thus the classifier would have to be evaluated with values that it has not seen during training.
- solution: a preprocessing step that accounts for the different types of transformations supported
- Whether the preprocessing step needs to be applied to a specific variable depends on the type of the variable and the transformation method utilized. 
- Table 1:  N/A indicates that the transformation method cannot be used for variables of the according to type.
- for all types of suppression (cell, attribute, record), random sampling as well as aggregation, evaluation data does not have to be preprocessed (the domain is preserved).
- all remaining transformation schemes, data need to be preprocessed before handing it to the classifier for evaluation. 
- preprocessing only needs to be performed for attribute values that have been generalized or categorized. In both cases,
- applying the same generalization hierarchies or categorization functions to input data that have also been used to anonymize the training dataset. 
- During the evaluation process this is performed automatically as all relevant information on how input data has been transformed is known to the software. 
- (notclear)For the purpose of utilizing the output data generated by ARX to build a privacy-preserving prediction model outside of the software, according to export functionalities (e.g. for hierarchies) are provided.
![Screenshot from 2022-03-14 16-22-33](https://user-images.githubusercontent.com/28802516/158204274-7393d5e7-d040-4a16-9fea-d2f941848009.png)

**Reply by idhamari** [reply]
**Performance assessment:**

* (what is dichotomous?)All implemented classification models are able to handle multinomial classification tasks, where the target variables need not be dichotomous. 
* Assessing the performance of multinomial classifiers is non-trivial and subject of ongoing research.
* One method to overcome this limitation is the one-vs-all approach, in which the performance of a n-nomial classifier is assessed by interpreting it as a collection of n binomial classifiers, each of which is able to distinguish one selected class from all others.
* supported measures: 
   - (1) sensitivity, also called recall or true positive rate. 
   - (2) Specificity, also called true negative rate. 
   - (3) The Receiver Operating Characteristic (ROC) curve, which plots the true positive rate (i.e. the sensitivity) for a single class against the false positive rate (1-specificity). The ROC curve shows the trade-off between sensitivity and specificity for every possible cutoff for a prediction, i.e. any increase in sensitivity will be accompanied by a decrease in specificity. 
   - (4) The Area Under the ROC Curve (ROC AUC), which summarizes the ROC performance of a classifier and which is equivalent to the probability that the classifier will assign a higher score to a randomly chosen positive event than to a randomly chosen negative event. 
   - (5) The Brier score: measures the mean squared distance between predicted and actual outcomes.
*  Zero Rule (0-R) algorithm, which ignores the feature variables and simply always returns the most frequent class value. 
* the performance of privacy-preserving models is reported in absolute terms as well as relative to baseline (0-R) and the selected classifier, both trained on unmodified input data.
* As an additional measure specific to our application scenario, we implemented the skill score, which quantifies the relative accuracy of a classification model over some reference accuracy. 
*  (conflict with the workshop example?) the relative accuracy is the accuracy of the classification model built from anonymized data over the accuracy of the model built from original data. 

![Screenshot from 2022-03-14 16-25-05](https://user-images.githubusercontent.com/28802516/158204767-9a382402-8ab3-40f9-81dd-5021286f549a.png)

* A skill score of zero means equality: that the Brier scores for models built on output and input data are equal. 
* If the score is in the range ] 0, 1] then the model built on output data performed better
* If it is in the range [ −∞, 0[, the model trained on the original data performed better.

**Comment by idhamari** [comment]
**Limitations and future work**

* Only supports three different types of prediction models, i.e. logistic regression, naïve Bayes and
random forest, for which we could find scalable implementations that are compatible to ARX in terms of their technical basis and licensing model. 
* Extend to C4.5 decision trees and support vector machines 
* Support benchmark studies for privacy-preserving models as well.
* supports transformation approaches where it is not clear how a given feature must be transformed to match the representation used for training purposes. Local generalization is an important example. develop more scalable approaches. 
* more learning and prediction tasks that are of specific importance to medical research. Important examples include regression and time-to-event analysis.
